### ISSUE TITLE
Refactor handling of shortcut-bots

### ISSUE THREAD

**Issue by Bibo-Joshi** [python-telegram-bot/python-telegram-bot#2505]
A bunch of TG-Objects have `bot` attribute that is used for the shortcuts, e.g. `self.bot.send_message(…)`, which is dangerous in the rare edge cases where users manually instantiate those objects and forget to pass the `bot`. This will then lead to somewhat unprecise error messages `NoneType has no attribute 'send_message'`.
Moreover, `Message.bot` is undocumented, but not private, so people relying on auto-complete may be using it without knowing that it's not really intended for external use.

## Proposal:

Do something like

```python
TGObject:
    self._bot: Optional[Bot] = None

    def get_bot(self) -> Bot:
        if self._bot is None:
            raise RuntimeError('This object has no bot and shortcuts cant be used')
        return self._bot

    def set_bot(self, bot: Bot) -> None:
        self._bot = bot
```

This way we have helpful error messages without having to check `if self.bot is None` in every shortcut.

An alternate solution would be to make `TGObject.bot` a property that raises the error if `_bot is None`. But explicitly calling a method makes it more clear that something *could* go wrong, IMO.

---
Note: We briefly discussed this in dev-chat on 04-05-2021, see https://t.me/c/1494805131/15168

**Comment by Poolitzer** [comment]
changing from `.bot` to `.get_bot()` could/should be considered breaking I guess, while using property setter wouldn't break anything.

I don't share the "more clear" mentality. Since its an undocumented property, raising this error is fine. Again, that this is even an problem happens only when someone outside of this library initiates their own objects, so they should then be fine with dealing with the error.


```suggestion
raise RuntimeError('This object was initiated without the bot parameter and therefore shortcuts can not be used')
```

**Comment by Bibo-Joshi** [comment]
> changing from `.bot` to `.get_bot()` could/should be considered breaking I guess, while using property setter wouldn't break anything.

True. Though I don't see a reason to make an effort to have this non-breaking as this is not urgent at all and we're headed for a breaking release anyway

> I don't share the "more clear" mentality. Since its an undocumented property, raising this error is fine. 

What I mean is: If I access an attribute of an object (that's either documented or non-private & suggested by auto-complete), I don't expect that to raise an exception in any case. I'm fine with checking that the value is something that I expect, but having to do

```python
try:
   obj.attr
except:
   print('accessing attr raised an exception')
```
is something that I would argue is unnatural.

> Again, that this is even an problem happens only when someone outside of this library initiates their own objects, so they should then be fine with dealing with the error.

Mh, my intention behind refactoring this has two aspects:

1. clearer error messages (something I remember you being fond of ;P)
2. making it clearer what users can rely on and what they can't. That's what I meant by
    > Moreover, Message.bot is undocumented, but not private, so people relying on auto-complete may be using it without knowing that it's not really intended for external use.

A different approach would be to just move `.bot` -> `._bot`. This would resolve the discrepancy "undocumented but not private". Whether or not we provide clearer error messages or not is then an independent decision.

> Also, I would rephrase the error:
> 
> ```
> raise RuntimeError('This object was initiated without the bot parameter and therefore shortcuts can not be used')
> ```

Sure, the thing above is just a quick draft.

**Comment by Poolitzer** [comment]
> making it clearer what users can rely on and what they can't.

So would you say `object.get_bot()` is clearer that it is private then `object.bot`?

Your try except block kind of convinced me though. It feels a bit weird.

**Comment by Bibo-Joshi** [comment]
> > making it clearer what users can rely on and what they can't.
> 
> So would you say `object.get_bot()` is clearer that it is private then `object.bot`?

No :D That part would be covered by a proper docstring in any of the discussed solutions (or making it private)

> Your try except block kind of convinced me though. It feels a bit weird.

🥳

**Comment by zpavloudis** [comment]
Hi all.
I've started working on this.
My initial approach is to attempt refactor of `self.bot ` as per this thread's comments and direction described in dev chat ( thanks @Poolitzer for providing a snippet of the conversation ). 

I've made changes to `Message.py` and `pytest/message.py/test_copy/` to attempt to validate my approach by running just this particular test. 

Do take a look at [zpavloudis/python-telegram-bot/commit/b980e7](https://github.com/zpavloudis/python-telegram-bot/commit/b980e7c84aaff9fa62d0f22ede6e72a5207f7e94). This currently fails throwing an AttributeError as tgRequest does not carry a post propert. 
```
>       orig_post = bot.request.post
E       AttributeError: 'property' object has no attribute 'post'
tests/conftest.py:713: AttributeError
```
Debugging indicates my approach is valid at least up to the point message.copy is called by the test.
That said I am new to this so would appreciate direction as to which TelegramObjects may require refactoring (all?) and any other thoughts on this would be highly regarded and appreciated.

Best,

**Comment by Bibo-Joshi** [comment]
Hi. Nice that you want to work on this! Your changes look good so far, only two things probably need some clarification:

1) For now we don't want to remove the possibility to pass the bot instance on initialization. So the parameter `bot` of `Message.__init__` should not get an underscore and [this line](https://github.com/zpavloudis/python-telegram-bot/blob/b980e7c84aaff9fa62d0f22ede6e72a5207f7e94/telegram/message.py#L550) should be changed to `self._bot = bot`
2) The `get/set_methods` you added are fine as they are, but not yet in the right space. The idea would be to add them to the class `TelegramObject`, such that all the base classes (like `Message`, `Chat`, `User`, …) automatically have these methods as well. Maybe adding them to `Message` was just for initial testing, but I wanted to comment on this, as there seemed to be have some confusion about this issue before :)

The reason for the failing test is the change in [this line](https://github.com/zpavloudis/python-telegram-bot/blob/b980e7c84aaff9fa62d0f22ede6e72a5207f7e94/tests/test_message.py#L1169), i.e. `message.set_bot(Bot)`. This sets `message._bot` to the _class_ `Bot` rather than an instance of that class. So when you pass `messag.get_bot()` to `check_defaults_handling` two lines below, you pass the class `Bot` and `Bot.request` is a property …

In the tests you actually don't have to worry about calling `set_bot` at all, because the `message` parameter is created such that a bot instance is passed to `Message.__init__` already (see [here](https://github.com/python-telegram-bot/python-telegram-bot/blob/bc7c422a11a60a3064601aac3fd5a26fd9b45ae9/tests/test_message.py#L240)). With 1. from above, you can just remove `message.set_bot(Bot)` :) (The same holds for all the other test files where we test shortcut methods)

Hopefully, this clarifies the goal of this issue a bit more. If you have any further questions, please don't hesitate to leave another comment here :)

**Comment by Bibo-Joshi** [comment]
Assigned you 👍
